### PR TITLE
Worksheet 3 - Group D

### PR DESCRIPTION
**Pull request by JoHaHu**

### CODE REVIEW

**Comment by JoHaHu** [comment]
The 2 failing pipeline steps are non-critical. The difference in the formatting is due to different versions of clang-format, and because we use pre-commit hooks to format it would be cumbersome to fix it. the clang-tidy issue is due to an code path that is already implemented and is working but is not used in the code path used in the simulations for the excercises from the sheet.

**Comment by JoHaHu** [comment]
Note it's important to use the specified compiler version, we use a lot of c++23 features and older compilers will fail

File: src/lib/container/linked_cell.h [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+      condition |= pos_y > bound_y;
+      break;
+  }
+  entry.active &= !condition;
+}
+
+/**
+ * function to apply reflecting boundary condition
+ * */
+
```
Comment (thesamriel) [inline src/lib/container/linked_cell.h:346]: These "main" functions would benefit from a more detailed documentation.

File: src/lib/container/linked_cell.h [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+template<index::Index I>
+static void reflecting(linked_cell<I> &lc, auto &p, orientation o, auto fc) {
+  auto [pos_x, pos_y, pos_z] = p.position;
+  auto [bound_x, bound_y, bound_z] = lc.index.boundary;
+  auto diff = lc.index.boundary - p.position;
+  auto [diff_x, diff_y, diff_z] = diff;
+
+  const double sigma = lc.sigma;
+  const double distance = std::pow(2, 1 / 6.0) * sigma;
```
Comment (thesamriel) [inline src/lib/container/linked_cell.h:355]: If I am not mistaken, this possibly expensive expression is calculated for every (boundary) particle.

File: test/LinkedCellsTest.cpp [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+  outflow_pc.boundary([](auto p) {});
+  outflow_pc.refresh();
+
+  auto counter = 0;
+  outflow_pc.linear([&counter](auto &p) {
+    counter++;
+  });
+  EXPECT_EQ(counter, 0);
+}
```
Comment (thesamriel) [inline test/LinkedCellsTest.cpp:142]: All of these tests could be combined into one test rather easily, which would avoid duplicating the code so many times.

**Review by thesamriel — APPROVED** [review]
**Note**
_This is only feedback regarding the code directly. Grade feedback will be on Moodle once we are through with all groups._

-----------------------------------------------------------------------------------------------------

Thanks a lot for your submission.
Your code has some really interesting approaches and usages of modern C++ features. I would personally say that your code is a little overengineered, and things are more complex than they need to be. But in the end, I also think that a lab course should be an open field to try out new things, as long as you fulfill the worksheet tasks. (Which you did)

You do many things manual, like memory allocation, iterating and so on. This can, when done right, have a positive impact on performance. However, it is also easy to mess things up and then things end up worse than a very basic implementation. I'm very curious to see how this will turn out.

One keypoint I want to mention is that you should not forget about 'basic' practices of efficient code. These include reordering of calculations, precomputing constant values or avoiding unnecessary copies.

As a last comment, I would appreciate if you could agree on one naming convention for classes and functions. You mix CamelCase and snake_case a lot, which makes reading the code even harder than it already is.

**Pull request merged**